I want sticker

theBaldSoprano opened the pull request:
hi @lanwen
added a method to testClass
and then made some test examples

theBaldSoprano committed:
add custom module

theBaldSoprano committed:
add UserSteps test

totallynotkate commented on commons-module/pom.xml:
```diff
@@ -0,0 +1,23 @@
+
+    <artifactId>commons-module</artifactId>
+
+    <dependencies>
```
У Apache в [примере](https://maven.apache.org/pom.html#Quick_Overview) pom.xml немного иначе структурирован: сначала идут dependencies, потом parent. Стоит придерживаться общепринятых конвенций, читать легче.

totallynotkate commented on commons-module/src/test/java/ru/qatools/school/MyFirstTest.java:
```diff
@@ -0,0 +1,19 @@
+ */
+public class MyFirstTest {
+
+    @Test
```
Если тест будет падать на первой проверке, он до второй не дойдет. Лучше разбить его на два независимых: один про Яндекс, другой про дом.

theBaldSoprano replied:
Согласен. Впихнул тут, чтобы посмотреть, что в обоих случаях Place работает верно.
Может ты сможешь ответить - почему мы создаем метод (в моем случае checkPlace) в классе ТЕСТИРУЕМОМ, а не в классе теста?

то, о чем ты говоришь "Если тест будет падать на первой проверке, он до второй не дойдет. " - это гипотетический вариант, который существовать и не сможет, ибо MyFirstTest - поле для моих проверок пока.
а вот assertThat(objectUnderTest, is(not(someOtherObject))); заюзать, т.к. наш метод существует под классом UserSteps и уже в нем реализован assert.

Вот тебе пример как я это вижу.
Покупаешь телевизор.
Что хочешь? - чтобы у тебя появился телевизор.
Что получаешь? - коробку с россыпухой и китайцем, который это все собирает при тебе.

:\ объясни плз

totallynotkate replied:
Неа, я про checkPlace ответить не могу, лучше спросить в чатике у Кирилла, наверное.
У меня MyFirstTest спокойно падал, когда я в UserSteps делала 

``` java
//this.place = place;
this.place = Place.AT_YANDEX;
```

В задании не запрещено добавить в UserSteps геттер, который тебе вернет место, и уже его запихнуть в assert, чтобы проверить, что ж там внутри телевизионной коробки.

lanwen replied:
Я чот не очень понял вашей дискуссии. Но попробую объяснить сакральный смысл проверочного метода с кишками вне теста и поля place.

Поле place - это имитация состояния системы (внешней). Просто сейчас оно хранится в поле. Обычно вы делаете какие то действия, чтобы состояние системы привести к нужному вам. Это делает метод `goTo`.

Дальше про проверочный метод. Вы сейчас не пишете приложение. Вы сейчас пишете тестовый проект. Который самодостаточен и написан специально для тестирования **другого** приложения. Соответственно бизнес логику взаимодействия с системой вы описываете в main - так чтобы ее можно было переиспользовать. Бизнес логика взаимодействия включает в себя как методы изменения системы, так и методы валидации состояния системы. Методы валидации в тестах должны использовать ассерты.

Тесты же сами (с аннотацией `@Test`) - это описания сценариев, по которым мы меняем стороннее приложение и валидируем его состояние. В тестах мы оперируем шагами. Т.е. если мы хотим проверить процесс включения телевизора после покупки, тест выглядит так: купили телевизор - привезли телевизор - включили - ждем что включился. Нам не нужно в тесте описывать что мы отдали в кассу столько то денег, что упаковали телевизор в коробку таких то размеров, довезли на газели с таким то номером, подняли на такой то этаж - нажали  на красную кнопку и ждем что экран замерцает с частотой 75 Гц. Это детали реализации, которые нам как пользователям не важны если мы просто хотим рабочий телевизор.

theBaldSoprano replied:
@totallynotkate @lanwen пример про телевизор был с такой задумкой:
использовать к тесте assert  - это как купить телевизор.
а использовать метод, в котором заюзан assert  - это как купить телевизор, а при этом к тебе домой приходит китаец с отвертками и деталями и собирает телек у вас на глазах, а потом еще и стоит в состоянии idle типа готовый его разобрать и собрать где-то в другом месте :)
теперь учитывая объясненное выше, я начал понимать почему нам интересен такой подход. и понятно это становится только если рассматривать тестовый проект как долгосрочный инструмент, дОлжный быть удобным тестеру. а так же, если учесть структуру - main с инструментами и test как бы наше рабочее пространство, где мы эти инструменты юзаем.  
типа у нас есть несколько методов под разные задачи, которые используют assert'ы и в тестах мы в зависимости от задачи вызываем нужный метод, который удобно описан в мэйн,а не херачим заново каждый раз конструкцию assert в тесте. 

я в правильную сторону мыслю? :)

lanwen replied:
да, все верно

totallynotkate commented on steps-module/src/main/java/ru/qatools/school/steps/UserSteps.java:
```diff
@@ -15,9 +18,14 @@ private UserSteps() {
     }
+
+    public void checkPlace(Place neededPlace) {
+        assertThat(place, is(neededPlace));
```
К проверкам стоит добавлять первым параметром строку с комментарием, что пошло не так.

theBaldSoprano committed:
- in UserSteps.java in method checkPlace rewrote assertThat with cool failcomment and used matchers for arguments to be more readable.
- in Place.java added one more variant for Place.*
- in MyFirstTest.java splitted existed @Test and added one more failing. all the work independently and dont fail each other

theBaldSoprano committed:
--it is an addition to prev commit(previously commited only one file for some reason. read on)

- in UserSteps.java in method checkPlace rewrote assertThat with cool failcomment and used matchers for arguments to be more readable.

- in Place.java added one more variant for Place.*

- in MyFirstTest.java splitted existed @Test and added one more failing. all the work independently and dont fail each other

- also corrected all pom'ses to be readable

theBaldSoprano commented:
@totallynotkate @lanwen  hi there Kate! (and Kirill if hes to view this ^^)
considered all your sayin and also added some thoughts of mine.
review plz, and all tips are appreciated/

totallynotkate commented on commons-module/src/test/java/ru/qatools/school/MyFirstTest.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Test
+    public void shouldBeAtHomeWhenGoHome() {
+        user().goTo(Place.HOME).checkPlace(Place.THE_MIDDLE_OF_NOWHERE);
```
Копипаста подвела, похоже, должно быть:

``` java
user().goTo(Place.HOME).checkPlace(Place.HOME);
```

theBaldSoprano replied:
неа @totallynotkate , специально завалил

lanwen replied:
если тест должен быть падающим намеренно, его стоит помечать как `@Ignore("причина + дата отключения + когда включить обратно или поменять")` - просто так падающий тест - бесполезен. На него все забивают, а заодно забивают и на остальные падающие тесты. Всем лень разбираться какого хрена падает тест если их больше 1.

totallynotkate replied:
Фуфуфу же, разработчики тебе будут в компот плевать за то, что тесты специально валишь. Я свои тоже пробую поломать, чтоб посмотреть, не проверяю ли я на самом деле true == true, а вся логика проходит мимо. Но в коммит потом идут чистенькие и красивые.

theBaldSoprano replied:
@lanwen @totallynotkate   спасибо! понял логику :+1:

theBaldSoprano replied:
@totallynotkate нихачу компот со слюнями :/

totallynotkate commented on steps-module/src/main/java/ru/qatools/school/data/Place.java:
```diff
@@ -5,5 +5,6 @@
     HOME,
-    AT_YANDEX
+    AT_YANDEX,
+    THE_MIDDLE_OF_NOWHERE
```
Не очень понятно, что тебе дает еще один элемент в перечислении. Расскажи, пожалуйста, зачем ты его добавил, мне любопытно)

theBaldSoprano replied:
@totallynotkate  "бог троицу любит"?)

totallynotkate replied:
Добавим в тесты немножечко магии :star2:

lanwen commented on steps-module/src/main/java/ru/qatools/school/steps/UserSteps.java:
```diff
@@ -15,9 +18,14 @@ private UserSteps() {
         return this;
     }
+
+    public void checkPlace(Place anExpectedPlace) {
```
слово `check` - плохое слово. Оно ничего не говорит кроме того что это "проверка". Но мы и так знаем что это проверка. Всегда лучше формулировать исходя из того что же тут должно быть. Т.е. "ждём", "должно быть"...

theBaldSoprano replied:
@lanwen не понимаю почему.

я мыслю так:
есть тест с говорящей фамилией
public void shouldBeAtYandexWhenGoToYandex() {
        user().goTo(Place.AT_YANDEX).checkPlace(Place.AT_YANDEX);
    }
должен быть в яндексе, когда пошел в яндекс.
ок
как мы проверим? -  воспользуемся методом "проверить/сравнить место"
checkPlace(Place anExpectedPlace)

разве не логично?

lanwen replied:
что значит "проверить"?

gladnik replied:
Помнишь слайд из [презенташки](https://speakerdeck.com/lanwen/ru-gimme-your-first-autotest), где `user().walksTo("Садовая").ordersCoffee().pays().shouldSeeCup()` (19-ый, там, кстати, ошибка, и можно потренироваться в её поисках)? 

Вот мы тут пишем такую же штуковину и хотим, чтобы она так же классно читалась. То есть название можно сделать, например, `shouldBeAtPlace()`, тогда в результате будет `user().goTo(null).shouldBeAtPlace(null)`.

lanwen replied:
вот вопрос в том что сам процесс "проверка" может быть с нюансами. А конкретное ожидение "должно быть таким же как" не допускает трактовок. В этом и прелесть слов should, expect... по сравнению с check.

_про ошибку и потренироваться в поисках_ не понял

gladnik replied:
Там на слайде аннотация `@Test` стоит не на месте, компилятор ругаться будет.

lanwen replied:
а. Там вообще нет метода, а выше - нет возвращаемых значений. Там псевдокод для понимания самого важного - внутренностей и структуры

lanwen commented on steps-module/src/main/java/ru/qatools/school/steps/UserSteps.java:
```diff
@@ -15,9 +18,14 @@ private UserSteps() {
     }
+
+    public void checkPlace(Place anExpectedPlace) {
+        assertThat("user must be in the same place that expected", this.place, is(anExpectedPlace));
```
длинные комментарии сложно поддерживать при изменении смысла проверки. Лучше быть более лаконичным и не повторять очевидных вещей. То что у нас что-то не такое как ожидаемое и так понятно. Здесь например, достаточно просто указать что это "неправильное место" - какое ожидалось и какое будет было получено - добавит матчер. 

Логично что всё должно быть в меру - если необходимо больше деталей - их нужно выводить. Пару раз поразбирав отчеты с кучей падений вы научитесь грамотно формулировать комментарии)

theBaldSoprano replied:
@lanwen  ty
